# Documentation

Genome indexing with Bowtie2

This wrapper takes the reference and build its index. The output files are 
stored in a directory extracted from the first output filename.

Required input:

- **reference**: the reference genome to index (FASTA format)

Required output:

- a list of expected output files. Need at least one. From the first file,
  the prefix is extracted and used by bowtie2 as **indexbase**.
 
Required parameters

- **options**: a list of valid fastqc options

Log:

- a log file generated by fastqc is stored

# Configuration

    #############################################################################
    #   MultiQC aggregates results from bioinformatics analyses across many
    #
    # :Parameters:
    #
    # - options: any options recognised by 'bowtie2 index' command
    # - threads: 
    bowtie2_index:
        options: ''
        threads: 4


# Example

    rule bowtie2_index:
        input:
            reference="genome.fa"
        output:
            "genome/bowtie2/genome.1.bt2"
        params:
            options = config['fastqc'][options'],
            working_directory = "{sample}/fastqc"
        threads:
            config['bowtie2_index']["threads"]            
        log:
            "logs/bowtie2_build/bowtie1_build.log"
        wrapper:
            "main/wrappers/bowtie2/build"

