# Documentation

Mapping sequencing reads onto a reference using Bowtie1

This wrapper takes single or paired data and map them on a reference. 
The results file is a sorted and indexed BAM file.

Required input:

- **fastq**: the input FastQ files (single or paired)
- **reference**: the reference genome to index (FASTA format)

Required output:

- **bam** :the output BAM file. 

Optional output (recommended):

- **sorted** If set, the BAM file will also be sorted and indexing provided
 
Required parameters

- **options**: a list of valid fastqc options
- **index** the expected index base of bowtie1 genome index

Log:

- a log file generated by fastqc is stored

# Configuration

    #############################################################################
    #  Bowtie1 read mapping
    #
    # :Parameters:
    #
    # - options: any options recognised by 'bowtie1 index' command
    # - threads: 
    bowtie1:
        options: ''
        threads: 4


# Example

    rule bowtie1:
        input:
            fastq=input_data
        output:
            bam="{sample}/bowtie1/{sample}.bam
            sorted="{sample}/bowtie1/{sample}.sorted.bam
        params:
            options = config['bowtie1'][options'],         
            index="reference/mygenome"
        threads:
            config['bowtie1"]["threads"]            
        log:
            "{sample}/bowtie1/{sample}.log"
        wrapper:
            "main/wrappers/bowtie1/align"

